Adding feature to ignore list and nuke errors

saurabh-hirani opened the pull request:
This PR adds the following features:

1. Support for --ignore-errors flag on command line as per the discussion in https://github.com/gruntwork-io/cloud-nuke/issues/112 

2. Updated ami.go to fix a bug which prints wrong number of deletions if ami deletion fails.

**Demo**:

1. Running cloud-nuke with `--ignore-errors all`:

```
cloud-nuke --ignore-errors all
.
.
INFO[2020-05-06T19:31:11+05:30] Checking region [1/1]: us-east-1
WARN[2020-05-06T19:31:12+05:30] Ignoring get resources error - asg - AccessDenied: User: arn:aws:sts::12345678:assumed-role/role-1/session-1 is not authorized to perform: autoscaling:DescribeAutoScalingGroups
        status code: 403, request id: xxxx
WARN[2020-05-06T19:31:12+05:30] Ignoring get resources error - lc - AccessDenied: User: arn:aws:sts::12345678:assumed-role/role-1/session-1 is not authorized to perform: autoscaling:DescribeLaunchConfigurations
        status code: 403, request id: xxxx
.
.
```
Complete command output - https://gist.github.com/saurabh-hirani/eaa9455488af5299ee5d07746a965f88 - does a best effort deletion. 

Detailed description as per the change in README.md

brikis98 commented:
Thanks for the PR! I'm super booked this week, but will try to take a look in the next few days.

brikis98 commented on README.md:
```diff
@@ -124,6 +124,29 @@ If you want to check what resources are going to be targeted without actually te
+
+Ideally you should have AWS IAM permissions to list and nuke all target resources. If you do not have permissions to
+nuke resource X - you should exclude it using the `--exclude-resource-type` flag. However, if you don't know which resource
+types you are not permitted to nuke or you have list acceess for all but nuke excess for some, you can ignore their
```
```suggestion
types you are not permitted to nuke or you have list access for all but nuke access for some, you can ignore their
```

saurabh-hirani replied:
Fixed.

brikis98 commented on README.md:
````diff
@@ -124,6 +124,29 @@ If you want to check what resources are going to be targeted without actually te
+cloud-nuke aws --ignore-errors ec2
+```
+
+This will ignore any errorrs encountered while listing/nuking EC2 resources - please note that if there are resources that
````
Please specify what other values you can pass here beyond `ec2` and how to handle multiple values.

brikis98 replied:
Also, does this ignore an EC2 error and continue nuking EC2 instances... Or does it stop nuking EC2 instances at the first EC2 error, but then keeps nuking other types of resources? We should make the behavior clear in the docs to avoid surprises.

saurabh-hirani replied:
Thanks for the feedback - clarified with example. It keeps on nuking other resource types - updated the README.

brikis98 commented on aws/aws.go:
```diff
@@ -481,8 +534,17 @@ func IsNukeable(resourceType string, resourceTypes []string) bool {
 
+// IsErrIgnorable - Checks if we should ignore errors for resource type or not
+func IsErrIgnorable(resourceType string, resourceTypes []string) bool {
+	if collections.ListContainsElement(resourceTypes, "all") ||
+		collections.ListContainsElement(resourceTypes, resourceType) {
+		return true
+	}
+	return false
```
```suggestion
	return collections.ListContainsElement(resourceTypes, "all") ||
		collections.ListContainsElement(resourceTypes, resourceType)
```

brikis98 commented on commands/cli.go:
```diff
@@ -135,27 +152,26 @@ func awsNuke(c *cli.Context) error {
-		}
-	}
+	// Check command line resource type values
+	invalidResourceTypes := map[string][]string{}
```
Why store these in a map?

saurabh-hirani replied:
Had kept them in a map in case we want to aggregate errors for list, exclude and ignore error types if user specifies invalid values. But there isn't a need of doing so at the moment - fixed it.

saurabh-hirani commented:
> Thanks for the PR!
> 
> Is there some way we can add automated tests to check this behavior?

I doubt if there is a straightforward way to do so. We are not checking for per resource type deletion errors - we are checking access related errors they return and deciding whether we want to move ahead or not. The way I tested it was to create a role that has permissions to create other roles - created a restricted role with it, applied that role's credentials to the cli, ran cloud-nuke and tried to nuke a resource not covered by the restricted role. 

Doing that in a test case would essentially be equal to retesting the above process. The way to introduce errors will lie not in the creation of the resource or its properties but in the role that tries to delete the resource. And to test if cloud-nuke goes ahead - we would essentially have to test multiple deletions while switching the creating and deleting roles. Does it make sense to do that? 

Honestly speaking - I was hoping this to be a feature which helps first time users get some best effort deletion work done and not fail if they encounter errors. I wouldn't use it if I know I have the right roles in place to run cloud-nuke.

brikis98 commented on README.md:
```diff
@@ -124,6 +124,35 @@ If you want to check what resources are going to be targeted without actually te
+### Ignoring list/nuke errors for certain/all resources
+
+Ideally you should have AWS IAM permissions to list and nuke all target resources. If you do not have permissions to
+nuke resource X - you should exclude it using the `--exclude-resource-type` flag. However, if you don't know which resource
```
```suggestion
nuke resource X, you should exclude it using the `--exclude-resource-type` flag. However, if you don't know which resource
```

brikis98 commented on README.md:
````diff
@@ -124,6 +124,35 @@ If you want to check what resources are going to be targeted without actually te
+cloud-nuke aws --ignore-errors ec2 --ignore-errors s3
+```
+
+This will ignore any errorrs encountered while listing/nuking EC2 and S3 resources. For example, if there are 3
````
```suggestion
This will ignore any errors encountered while listing/nuking EC2 and S3 resources. For example, if there are 3
```

brikis98 commented on README.md:
````diff
@@ -124,6 +124,35 @@ If you want to check what resources are going to be targeted without actually te
+```
+
+This will ignore any errorrs encountered while listing/nuking EC2 and S3 resources. For example, if there are 3
+EC2 instances and 2 S3 buckets to nuke and few/all of the EC2 instances and few/all of the S3 bucket deletion fails - cloud-nuke
````
```suggestion
EC2 instances and 2 S3 buckets to nuke and few/all of the EC2 instances and few/all of the S3 bucket deletion fails, cloud-nuke
```

brikis98 commented on README.md:
```diff
@@ -124,6 +124,35 @@ If you want to check what resources are going to be targeted without actually te
+EC2 instances and 2 S3 buckets to nuke and few/all of the EC2 instances and few/all of the S3 bucket deletion fails - cloud-nuke
+will log those errors and proceed with deleting further resources.
+
+Please note that cloud-nuke deletes resource types in a certain order to avoid dependency errors. If there are resources
+that should be deleted before EC2 or S3 - and there are errors encountered while listing/nuking those resources e.g.
+cloud-nuke deletes autoscaling groups before EC2 and if your role does not have permissions to delete autoscaling groups
+and you have not added `asg` to the `--ignore-errors` list, then cloud-nuke will error out and fail before reaching to EC2.
```
```suggestion
Please note that cloud-nuke deletes resource types in a certain order to avoid dependency errors. If there are resources
that should be deleted before EC2 or S3, and there are errors encountered while listing/nuking those resources, then cloud-nuke will error out and fail before reaching to EC2 or S3. For example,
cloud-nuke deletes autoscaling groups before EC2 instances, and if your role does not have permissions to delete autoscaling groups
and you added `ec2` to the `--ignore-errors` list but not `asg`, then cloud-nuke will still end up exiting with an error around deleting those autoscaling groups.
```

brikis98 commented on README.md:
````diff
@@ -124,6 +124,35 @@ If you want to check what resources are going to be targeted without actually te
+errors by specifying the `--ignore-errors` flag:
+
+```shell
+cloud-nuke aws --ignore-errors ec2 --ignore-errors s3
````
We still need to provide users with the a pointer to what resource types can be specified via `--ignore-errors`. i.e., It's more than just `ec2` and `s3`, right?

brikis98 commented:
> Doing that in a test case would essentially be equal to retesting the above process. The way to introduce errors will lie not in the creation of the resource or its properties but in the role that tries to delete the resource. And to test if cloud-nuke goes ahead - we would essentially have to test multiple deletions while switching the creating and deleting roles. Seems like a bigger overhead than the value it provides. Does it make sense to do that?

I actually think it does make sense to test this! In particular, I'd create automated tests for two scenarios:

1. Create an IAM role that _only_ has read permissions for EC2 and S3. Create an S3 bucket and an EC2 instance, assume the role, run `cloud-nuke` telling it to ignore errors, and make sure the run completes successfully, and that the S3 bucket and EC2 instance are still around (which the test can then clean up at the end).
1. Same as above, except this time, the IAM role also has write permissions for S3, so at the end of the test, the S3 bucket should be deleted, but the EC2 instance should still be around.

saurabh-hirani commented:
> 1. Create an IAM role that _only_ has read permissions for EC2 and S3. Create an S3 bucket and an EC2 instance, assume the role, run `cloud-nuke` telling it to ignore errors, and make sure the run completes successfully, and that the S3 bucket and EC2 instance are still around (which the test can then clean up at the end).
> 2. Same as above, except this time, the IAM role also has write permissions for S3, so at the end of the test, the S3 bucket should be deleted, but the EC2 instance should still be around.

Okay - will give it a shot and update this thread with findings.

saurabh-hirani commented:
@brikis98 - am able to try out the individual functions for creating/assuming role and 
creating/deleting policies for the same. I had the following queries before proceeding:

1. Having a file called `cli_nukeallresources_test.go` (more apparent in later queries)
besdies `cli_test.go`. But as we have to create test EC2 and S3 instances - can we reuse the
corresponding `createTestEC2Instance` in `aws/ec2_test.go` - that would mean uppercasing that func to export it and upper casing all its dependent funcs. `getRandomRegion` cannot be renamed as it is heavily used outside also. Or does it make sense to repeat the test EC2 creation code in the test case to avoid any dependencies? 

Same query for creating test S3 buckets but S3 is more complicated as its PR is still 
open - https://github.com/gruntwork-io/cloud-nuke/pull/110 

2. Having the following test functions in `cli_nukeallresources_test.go` - does this coverage
make sense?

- TestNukeAllResources_AllPerms() 
    - Creates test EC2, S3
    - No assume role
    - Call GetAllRresources - for resource type - EC2, S3 - returns list with test EC2 and S3
    - Call NukeAllResources with test EC2 and S3 - deletes them
    - Test EC2, S3 do not exist

- TestNukeAllResources_NoPerms()
    - Creates test EC2, S3
    - Creates test role with no read/write for EC2, S3 
    - Assume role
    - Call GetAllRresources - for resource type - EC2, S3 - returns errors
    - Test EC2, S3 still exist

- TestNukeAllResources_NoPerms_IgnoreErrors()
    - Same as above but GetAllResurces call has ignore errors flag set - so it doesn't fail  and returns empty list.
    - Test EC2, S3 still exist

- TestNukeAllResources_EC2S3_RORole()
    - Creates test EC2, S3
    - Creates test role with read for EC2, S3 
    - Assume role
    - Call GetAllRresources - for resource type - EC2, S3 - no errors - returns test EC2 and S3
    - Call NukeAllResources with test EC2 and S3 - returns errors
    - Test EC2, S3 still exist

- TestNukeAllResources_EC2S3_RORole_IgnoreErorors()
    - Same as above but NukeAllResources has ignore errors flag set - so it doesn't fail  and returns nil.
    - Test EC2, S3 still exist

- TestNukeAllResources_EC2_RORole_S3_RWRole()
    - Creates test EC2, S3
    - Creates test role with read for EC2, write for S3 
    - Assume role
    - Call GetAllRresources - for resource type - EC2, S3 - no errors - returns test EC2 and S3
    - Call NukeAllResources with test EC2 and S3 - fails at EC2 - does not go upto S3
    - Test EC2, S3 still exist

- TestNukeAllResources_EC2_RORole_S3_RWRole_IgnoreErrors()
    - Same as above but NukeAllResources has ignore errors flag set - so it doesn't fail 
        for EC2 and continues S3 deletion and returns nil.
    - Test S3 exists. Test EC2 deleted.

brikis98 commented:
> Having a file called cli_nukeallresources_test.go (content fleshed out in the next query)
besides cli_test.go. But as we have to create test EC2 and S3 instances - can we reuse the
corresponding createTestEC2Instance in aws/ec2_test.go - that would mean uppercasing that func to export it. getRandomRegion cannot be renamed as it is heavily used outside also. Or does it make sense to repeat the test EC2 creation code in the test case to avoid any dependencies?

Feel free to move these functions to some package (e.g.,`test_util`) so they can be imported wherever needed.

> Having the following test functions in cli_nukeallresources_test.go - does this coverage
make sense?

That test coverage sounds great 👍 

One NIT: Go naming convention is CamelCase and not snake_case or Upper_Snake_case. So the function names should be `TestNukeAllResourcesEC2S3ReadOnlyRoleIgnoreErorors`... But otherwise, LGTM!

saurabh-hirani commented:
@brikis98 Updated as per feedback. As the config file PR also touched some common files - `cli.go`, `aws.go`, `s3_test.go` - had to rebase and the diffs didn't seem reviewable. So squashed the commits to review as one unit. Please review and let me know if this looks good. 

Summary:

- Added the following new files in `aws/`:
  - `ec2_utils.go` - moving out common test CRUD functions for ec2 - led to updating test files - `ec2_test.go` and dependent test files - `ebs_test.go, asg_test.go, ami_test.go, launch_config_test.go` .
  - `s3_utils.go` - same as above but for S3 - led to updating test files - `s3_test.go`
  - `iam_utils.go` - new file for CRUD for IAM roles/policies - doubles up as a base for future addition of IAM role/policy nuking

- Added the following new files in `commands/`:
  - `cli_nukeallresources_test.go` - contains test cases as described above - basically uses ec2/s3 utils for creating/reading/deleting resources and iam utils for creating/assuming/deleting roles. 

- Updated `aws.GetAllResources` and `aws.NukeAllResources` to accept session param because in order to assume role you have to pass in session param while deleting resources - which wasn't there. The argument list of these 2 functions was getting longer and hence used a struct to make it simpler + support optional args.

- Updated `ec2.go` to support flag which decides whether to wait or not for instance deletion after triggering it - https://github.com/gruntwork-io/cloud-nuke/pull/114/files#diff-6c5db3fd88049ee01a36ba713dfcdbdeR76 - Ec2 nuke to not wait for instance termination by default - default case is trigger and return. This is because when testing deletion repeatedly with around 8-10 EC2 instances - (encountered when deferred nukes run for assume role tests) - there were random timeouts + long delay sometimes for EC2 deletion. For dependent tests like `ebs_test.go` they wait for EC2 deletion e.g https://github.com/gruntwork-io/cloud-nuke/pull/114/files#diff-c57ef596e24478400f32b6312d13256eR183

- There is no support to pass in trigger-delete-and-return/wait-for-delete-complete of cloud-nuke resources. Can be taken up as separate feature request. 

- When running `s3_test.go` - the following tests failed for me - because as per https://github.com/gruntwork-io/cloud-nuke/pull/113#issuecomment-643485861 some test run in Gruntwork specific accounts e.g. phxdevops - and I am running them in my own account. Wanted to check if the corresponding tests should be rewritten to create test buckets so that they can run on non-Gruntwork AWS accounts also. 

```
--- FAIL: TestFilterS3Bucket_Config (30.65s)
    --- FAIL: TestFilterS3Bucket_Config/config_tests (0.00s)
        --- FAIL: TestFilterS3Bucket_Config/config_tests/Include (0.20s)
            s3_test.go:441:
                        Error Trace:    s3_test.go:441
                        Error:          Not equal:
                                        expected: 4
                                        actual  : 0
                        Test:           TestFilterS3Bucket_Config/config_tests/Include
        --- FAIL: TestFilterS3Bucket_Config/config_tests/IncludeAndExclude (0.22s)
            s3_test.go:441:
                        Error Trace:    s3_test.go:441
                        Error:          Not equal:
                                        expected: 3
                                        actual  : 0
                        Test:           TestFilterS3Bucket_Config/config_tests/IncludeAndExclude
        --- FAIL: TestFilterS3Bucket_Config/config_tests/Exclude (0.27s)
            s3_test.go:441:
                        Error Trace:    s3_test.go:441
                        Error:          Not equal:
                                        expected: 6
                                        actual  : 0
                        Test:           TestFilterS3Bucket_Config/config_tests/Exclude
```

saurabh-hirani commented:
@brikis98 @yorinasub17  This change has more testing code than the actual functionality :) - maybe the tests required very methodical series of steps (create role, assume role, create resources, try deletion, delete resources, delete role) to verify ignoring errors.

saurabh-hirani commented:
@brikis98 - can you please review this and check if this looks good? Thanks.

brikis98 commented:
Sorry for the delay. It's a huge PR, so I keep snoozing it every time I see it 😁 

I'll try to get to it in the next week or so!

brikis98 commented on commands/cli_nukeallresources_test.go:
```diff
@@ -0,0 +1,547 @@
+	}
+
+	// Assume role
+	var roleSession *session.Session
+	if assumeRoleArn == "" {
+		roleSession = awsParams.AWSSession
+	} else {
+		roleSession, err = aws.AssumeIAMRole(assumeRoleArn, awsParams.Region)
+		require.NoErrorf(t, err, "Failed to assume role - %s - %s", assumeRoleArn, err)
+	}
```
Slightly simpler:

```suggestion
	roleSession := awsParams.AWSSession
	if assumeRoleArn != "" {
		roleSession, err = aws.AssumeIAMRole(assumeRoleArn, awsParams.Region)
		require.NoErrorf(t, err, "Failed to assume role - %s - %s", assumeRoleArn, err)
	}
```

brikis98 commented on commands/cli_nukeallresources_test.go:
```diff
@@ -0,0 +1,547 @@
+	logging.Logger.Debug("Expected: GetAllResources passed")
+
+	// Validate that get call should return the right resource types
+	if collections.ListContains(args.getResourcesOutputTypes, "ec2") {
+		require.Truef(
+			t, resourceExists(awsParams.Region, "ec2", resourceTypeIdentifierMap["ec2"], account),
+			"Failed to find test EC2 instance - name - %s - instanceID - %s", ec2InstanceName, resourceTypeIdentifierMap["ec2"],
+		)
+		logging.Logger.Debug("Expected: Post get check: ec2 exists")
+	}
+	if collections.ListContains(args.getResourcesOutputTypes, "s3") {
+		require.Truef(
+			t, resourceExists(awsParams.Region, "s3", bucketName, account),
+			"Failed to find test S3 bucket - %s", bucketName,
+		)
+		logging.Logger.Debug("Expected: Post get check: s3 exists")
+	}
```
Seems like you could loop over `args.getResourceOutputTypes` and do this check dynamically, rather than hard-coding ec2 and s3 checks?

brikis98 commented on commands/cli_nukeallresources_test.go:
```diff
@@ -0,0 +1,547 @@
+	)
+	// Validate that get call should fail if specified
+	if args.getResourcesShouldFail {
+		require.Error(t, err, "GetAllResources did not fail")
```
Hm, prob a good idea to check _which_ error you get here!

saurabh-hirani replied:
Given the current codebase, I don't think we can check. For example `GetAllResources` calls get for all resource types, each of which returns `errors.WithStackTrace(err)` - how do we check specific error type in this case?

brikis98 commented on commands/cli_nukeallresources_test.go:
```diff
@@ -0,0 +1,547 @@
+
+	// Validate that nuke call should fail if specified
+	if args.nukeResourcesShouldFail {
+		require.Error(t, err, "assumedRole NukeAllResources did not fail")
```
Same here

saurabh-hirani replied:
`NukeAllResources` has the same situation as `GetAllResources`

brikis98 commented on commands/cli_nukeallresources_test.go:
```diff
@@ -0,0 +1,547 @@
+	logging.Logger.Debug("Expected: GetAllResources passed")
+
+	// Validate post nuke resources
+	ec2Exists := resourceExists(awsParams.Region, "ec2", resourceTypeIdentifierMap["ec2"], account)
+	s3Exists := resourceExists(awsParams.Region, "s3", resourceTypeIdentifierMap["s3"], account)
+
+	if collections.ListContains(args.postNukeResources, "ec2") {
+		require.Truef(
+			t, ec2Exists,
+			"Test EC2 instance deleted when it should not have - name - %s - instanceID - %s",
+			ec2InstanceName, resourceTypeIdentifierMap["ec2"],
+		)
+		logging.Logger.Debug("Expected: Post nuke check - ec2 exists")
+	} else {
+		require.Falsef(
+			t, ec2Exists,
+			"Test EC2 instance not deleted - name - %s - instanceId - %s",
+			ec2InstanceName, resourceTypeIdentifierMap["ec2"],
+		)
+		logging.Logger.Debug("Expected: Post nuke check - ec2 deleted")
+	}
+	if collections.ListContains(args.postNukeResources, "s3") {
+		require.Truef(t, s3Exists, "Test S3 bucket deleted when it should not have - %s", bucketName)
+		logging.Logger.Debug("Expected: Post nuke check - s3 exists")
+	} else {
+		require.Falsef(t, s3Exists, "Test S3 bucket not deleted - %s", bucketName)
+		logging.Logger.Debug("Expected: Post nuke check - s3 deleted")
+	}
```
This could prob be done dynamically in a loop too?

brikis98 commented on commands/cli_nukeallresources_test.go:
```diff
@@ -0,0 +1,547 @@
+	}
+}
+
+// TestNukeAllResourcesNoPerms tests deletion of test S3 and EC2 instance with an assumed
+// role which does not have S3/EC2 read/write permissions
```
```suggestion
// TestNukeAllResourcesNoPerms tests deletion of test S3 buckets and EC2 instances with 
// various IAM permissions: no access, read-only access, read/write access for some but 
// read-only for others, full read/write access.
```

brikis98 commented on commands/cli_nukeallresources_test.go:
```diff
@@ -0,0 +1,547 @@
+func TestNukeAllResources(t *testing.T) {
+	t.Parallel()
+	// Create a top level AWS session object which will be used to create/destroy roles
+	// and resources. Specifically use us-east-1 as we are running >= 8 tests and vCpu limits
```
Are the default limits for `us-east-1` any higher? Or is it just that you requested limited increases already for that region?

saurabh-hirani replied:
These are the default limits. I did not request an increase.

brikis98 commented on commands/cli_nukeallresources_test.go:
```diff
@@ -0,0 +1,547 @@
+		assumeRolePolicyDocumentTmpl, "$callerIdentityArn", callerIdentityArn, -1,
+	)
+
+	var testCases = []struct {
```
Very nice table-driven tests 👍

brikis98 commented on commands/cli_nukeallresources_test.go:
```diff
@@ -0,0 +1,547 @@
+		{
+			"NoPerms",
+			createIAMRoleArgs{
+				roleName:                 "cloud-nuke-test-1-noperms",
```
Does this name get used exactly as-is? If so, this test may fail due to a name conflict if we have two commits that kick off the test suite in parallel. Could you ensure that the role name has a unique ID appended to it using `util.UniqueId()`.

saurabh-hirani replied:
Makes sense - will do.

brikis98 commented on commands/cli_nukeallresources_test.go:
```diff
@@ -0,0 +1,547 @@
+				assumeRolePolicyDocument: assumeRolePolicyDocument,
+			},
+			createIAMRolePolicyArgs{
+				roleName:       "cloud-nuke-test-1-noperms",
```
Why does the policy need a role name?

saurabh-hirani replied:
Because it is an inline policy attached to a role -https://docs.aws.amazon.com/IAM/latest/APIReference/API_PutRolePolicy.html

brikis98 commented on commands/cli_nukeallresources_test.go:
```diff
@@ -0,0 +1,547 @@
+}
+
+// getIAMRolePolicyMamp maps a role type to a corresponding policy
+func getTestRolePolicyMap() map[string]string {
```
Why use a map with string-based key lookup instead of a few named constants?

saurabh-hirani commented on aws/ami_test.go:
```diff
@@ -33,7 +33,10 @@ func waitUntilImageAvailable(svc *ec2.EC2, input *ec2.DescribeImagesInput) error
 	svc := ec2.New(session)
-	instance := createTestEC2Instance(t, session, name, false)
+	instance, err := CreateTestEC2Instance(session, name, false)
+	if err != nil {
+		assert.Failf(t, "Could not create test EC2 instance", errors.WithStackTrace(err).Error())
```
Have done that in the new files I added, but didn't do so in the old ones as they all were using `assert.Fail` and wanted to keep in sync. Ideally they should be using `require` where we want the test to not proceed further. But didn't want to do that level of yak shaving. Is that ok or should I use `require` for all the new code I added?

saurabh-hirani commented on aws/ami_test.go:
```diff
@@ -104,7 +107,11 @@ func TestListAMIs(t *testing.T) {
 	// clean up after this test
 	defer nukeAllAMIs(session, []*string{image.ImageId})
-	defer nukeAllEc2Instances(session, findEC2InstancesByNameTag(t, session, uniqueTestID))
+	instances, err := findEC2InstancesByNameTag(session, uniqueTestID)
+	if err != nil {
+		assert.Fail(t, errors.WithStackTrace(err).Error())
+	}
```
Used to checking `err` immediately after the call but I see your point. Would changing

```

	// clean up after this test
	defer nukeAllAMIs(session, []*string{image.ImageId})
	instances, err := findEC2InstancesByNameTag(session, uniqueTestID)
	if err != nil {
		assert.Fail(t, errors.WithStackTrace(err).Error())
	}
	defer nukeAllEc2Instances(session, instances, true)
```

to 

```

	// clean up after this test
	defer nukeAllAMIs(session, []*string{image.ImageId})
	instances, err := findEC2InstancesByNameTag(session, uniqueTestID)
	defer nukeAllEc2Instances(session, instances, true)
         if err != nil {
		assert.Fail(t, errors.WithStackTrace(err).Error())
	}
```

make sense?

saurabh-hirani commented on aws/ec2_utils.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+// CreateTestEC2Instance - creates a test EC2 instance
+func CreateTestEC2Instance(session *session.Session, name string, protected bool) (ec2.Instance, error) {
```
This is no longer in `ec2_test.go` as it is moved to `ec2_utils.go` because `cli_nukeallresources_test.go` creates test S3 and test EC2 instance and we needed that logic to be moved out of test files as per. earlier discssion. So I moved the corresponding utility functions in `ec2_test.go` and `s3_test.go`. Going by the normal convention of creating an AWS resource or returning error if there is a failure. Cannot do an `assert` or a `require` in this method. Does that make sense or is there a better way around this?

saurabh-hirani commented on aws/ec2.go:
```diff
@@ -92,6 +93,11 @@ func nukeAllEc2Instances(session *session.Session, instanceIds []*string) error
 		return errors.WithStackTrace(err)
 	}
 
+	if !waitForTermination {
```
Had added it to support async deletion. Reverted it back as it might be better tackled in a separate issue. Sometimes, we just want to trigger deletion and not wait for resources that take a long time to nuke.

saurabh-hirani commented on aws/iam_utils.go:
```diff
@@ -0,0 +1,236 @@
+	"github.com/gruntwork-io/cloud-nuke/logging"
+)
+
+func createSessionIfNotExists(awsSession *session.Session) *session.Session {
+	if awsSession == nil {
```
Not required - this was to prevent creating session in the `cli_testnukeresources.go` but it is better done there and this doesn't read well.

saurabh-hirani commented on aws/iam_utils.go:
```diff
@@ -0,0 +1,236 @@
+	return err
+}
+
+// ValidateIAMRolePolicy validates an IAM role inline policy
+func ValidateIAMRolePolicy(awsSession *session.Session, roleName string, policyName string) error {
```
Thanks for catching this - it was added earlier - not required as it is not being called. Will fix.

saurabh-hirani commented on aws/iam_utils.go:
```diff
@@ -0,0 +1,236 @@
+}
+
+// DeleteIAMRolePolicy deletes IAM role policy
+func DeleteIAMRolePolicy(awsSession *session.Session, roleName string, policyName string) error {
```
It should. Will fix.

saurabh-hirani commented on aws/iam_utils.go:
```diff
@@ -0,0 +1,236 @@
+	return sess, nil
+}
+
+// TrimPolicyDocument converts newlines and spaces into a policyDocument format
+// required while creating role
```
Yes. Will fix.

saurabh-hirani commented on commands/cli_nukeallresources_test.go:
```diff
@@ -0,0 +1,547 @@
+			continue
+		}
+		for _, resourceCollectionObj := range accountResources.Resources {
+			if resourceCollectionObj.ResourceName() == targetResourceType {
```
Short answer: `ResourceName()` is a misnomer - it actually points to a resource's type. 


Long answer:

`accountResources` is of type 

```
type AwsAccountResources struct {
	Resources map[string]AwsRegionResource
}
```

where AwsRegionResource is 

```
type AwsRegionResource struct {
	Resources []AwsResources
}
```

and AwsResources is 

```
type AwsResources interface {
	ResourceName() string
	ResourceIdentifiers() []string
	MaxBatchSize() int
	Nuke(session *session.Session, identifiers []string) error
}
```

The `ResourceName` key in this type should ideally be called `ResourceType` because ResourceNames key in this is actually `ResourceIdentifiers`. All resources use `ResourceName` when they should mean `ResourceType` e.g. 

```

// ResourceName - the simple name of the aws resource
func (instance EC2Instances) ResourceName() string {
        return "ec2"
}
```

```

// ResourceName - the simple name of the aws resource
func (address EIPAddresses) ResourceName() string {
        return "eip"
}
```

saurabh-hirani commented:
Thanks for the detailed feedback @brikis98 - I appreciate that. I have closed most of them and have put in queries where I needed some guidance. Smoke tests on `cli_nukeallresources_test.go` look good - will be testing other `_test.go` files tomorrow and will update this thread.

saurabh-hirani commented:
@brikis98 as per above ran tests for the following test files also - looks good 

```
ami_test.go
asg_test.go
ebs_test.go
ec2_test.go
s3_test.go
```

saurabh-hirani commented:
We can close this PR as it's review has been pending for a few months and it has deviated from the main codebase.
